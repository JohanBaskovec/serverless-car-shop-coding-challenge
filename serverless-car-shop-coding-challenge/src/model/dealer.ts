import mongoose, { Schema } from 'mongoose';
import { type DealerDocument, type DealerModel, type DealerSchema } from '../generated/mongoose.gen.js';

export const dealerSchema: DealerSchema = new Schema<DealerDocument>({
    name: {
        type: String,
        required: true
    },
    vehicles: [{
        type: Schema.Types.ObjectId,
        ref: 'Vehicle',
        required: true
    }]
}, {
    timestamps: {
        createdAt: 'creationDate',
        updatedAt: 'lastUpdateDate'
    }
});

export const Dealer: DealerModel = mongoose.model<DealerDocument, DealerModel>('Dealer', dealerSchema);
