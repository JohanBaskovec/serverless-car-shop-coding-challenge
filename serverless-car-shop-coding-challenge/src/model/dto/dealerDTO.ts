import { type components } from '../../generated/openapi.js';

export type CreateDealerDTO = components['schemas']['CreateDealerDTO'];
export type UpdateDealerDTO = components['schemas']['UpdateDealerDTO'];
