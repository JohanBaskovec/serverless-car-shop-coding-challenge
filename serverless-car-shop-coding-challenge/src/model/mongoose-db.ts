import { connect, type Mongoose } from 'mongoose';

export async function connectToMongo(): Promise<Mongoose> {
    if (process.env.DB_URL == null) {
        throw new Error('process.env.DB_URL must be defined.');
    }
    return await connect(process.env.DB_URL);
}
