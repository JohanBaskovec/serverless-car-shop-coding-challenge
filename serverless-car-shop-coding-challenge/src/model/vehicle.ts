import mongoose, { Schema } from 'mongoose';
import { type VehicleDocument, type VehicleModel, type VehicleSchema } from '../generated/mongoose.gen';
import { type components } from '../generated/openapi.js';

export type FuelType = components['schemas']['FuelType'];

export type VehicleType = components['schemas']['VehicleType'];

export const vehicleSchema: VehicleSchema = new Schema<VehicleDocument>({
    make: {
        type: String,
        required: true
    },
    // naming the property "model" makes it impossible to mock the module
    vehicleModel: {
        type: String,
        required: true
    },
    transmission: {
        type: String,
        required: true,
        enum: [
            'manual',
            'automatic',
            'semiAutomatic'
        ]
    },
    fuelType: {
        type: String,
        required: true,
        enum: [
            'petrol',
            'diesel',
            'eletric',
            'lpg',
            'hybrid'
        ] as FuelType
    },
    mileage: {
        type: Number,
        required: true
    },
    vehicleType: {
        type: String,
        required: true,
        enum: [
            'cabriolet',
            'coupe',
            'estateCar',
            'suv',
            'saloon',
            'van',
            'smallCar',
            'other'
        ] as VehicleType
    },
    color: {
        type: String,
        required: true
    },
    dealer: {
        type: Schema.Types.ObjectId,
        ref: 'Dealer',
        required: true
    }
}, {
    timestamps: {
        createdAt: 'creationDate',
        updatedAt: 'lastUpdateDate'
    }
});

export const Vehicle: VehicleModel = mongoose.model<VehicleDocument, VehicleModel>('Vehicle', vehicleSchema);
