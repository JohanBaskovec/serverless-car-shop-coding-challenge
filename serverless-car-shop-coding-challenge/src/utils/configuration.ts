import * as dotenv from 'dotenv';
import * as path from 'node:path';

export function initializeEnvironmentVariables(): void {
    const env = process.env.CODING_CHALLENGE_ENV ?? 'dev';
    const dotenvPath = path.join(__dirname, `../../config/.env.${env}`);
    dotenv.config({
        path: dotenvPath
    });
}
