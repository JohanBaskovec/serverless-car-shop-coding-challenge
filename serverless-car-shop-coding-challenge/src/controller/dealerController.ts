import { type CreateDealerDTO, type UpdateDealerDTO } from '../model/dto/dealerDTO.js';
import { CrudController } from './crudController.js';
import { type DealerDocument } from '../generated/mongoose.gen.js';

export class DealerController extends CrudController<DealerDocument, CreateDealerDTO, UpdateDealerDTO> {
}
