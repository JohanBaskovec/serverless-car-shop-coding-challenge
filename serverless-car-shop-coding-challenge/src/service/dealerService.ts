import { Dealer } from '../model/dealer.js';
import { type CreateDealerDTO, type UpdateDealerDTO } from '../model/dto/dealerDTO.js';
import { ConstraintError, EntityDoesNotExistError, MissingKeyInJSONBodyError } from '../error/errors.js';
import mongoose from 'mongoose';
import { type VehicleService } from './vehicleService.js';
import { type DealerDocument, type VehicleDocument } from '../generated/mongoose.gen.js';
import { type CrudService } from './crudService.js';
import { type APIGatewayProxyEventQueryStringParameters } from 'aws-lambda/trigger/api-gateway-proxy.js';
import { type DeleteResponse } from '../model/response.js';

export class DealerService implements CrudService<DealerDocument, CreateDealerDTO, UpdateDealerDTO> {
    private _vehicleService: VehicleService;

    get vehicleService(): VehicleService {
        return this._vehicleService;
    }

    set vehicleService(value: VehicleService) {
        this._vehicleService = value;
    }

    async getAll(queryStringParameters: APIGatewayProxyEventQueryStringParameters | null = null): Promise<DealerDocument[]> {
        return await Dealer.find({}).exec();
    }

    async create(dto: CreateDealerDTO): Promise<DealerDocument> {
        const dealer = await new Dealer(dto).save();
        for (const vehicle of dealer.vehicles) {
            await this.vehicleService.setDealer(vehicle, dealer, false);
        }
        return dealer;
    }

    async update(dto: UpdateDealerDTO): Promise<DealerDocument> {
        if (dto._id == null) {
            throw new MissingKeyInJSONBodyError(['_id']);
        }
        if (dto.vehicles != null) {
            await this.setVehicles(
                new mongoose.Types.ObjectId(dto._id),
                dto.vehicles.map(v => new mongoose.Types.ObjectId(v)),
                true
            );
        }
        delete dto.vehicles;
        const dealer = await Dealer.findOneAndUpdate({ _id: dto._id }, dto, { new: true, runValidators: true }).exec();
        if (dealer == null) {
            throw new EntityDoesNotExistError(`Dealer with id ${dto._id} does not exist.`);
        }
        return dealer;
    }

    async remove(id: string): Promise<DeleteResponse> {
        const dealer: DealerDocument | null = await Dealer.findOne({ _id: id }).exec();
        if (dealer == null) {
            throw new EntityDoesNotExistError(`Dealer with id ${id} does not exist.`);
        }
        if ((dealer.vehicles as mongoose.Types.ObjectId[]).length !== 0) {
            throw new ConstraintError('Dealer has vehicles, delete them first before deleting the dealer.');
        }
        const response = await Dealer.deleteOne({ _id: id }).exec();
        if (response.deletedCount === 0) {
            throw new EntityDoesNotExistError(`Dealer with id ${id} does not exist.`);
        }
        return response;
    }

    async removeVehicle(
        dealerOrId: DealerDocument | mongoose.Types.ObjectId,
        vehicle: mongoose.Types.ObjectId | VehicleDocument
    ): Promise<void> {
        const dealer = await this.getDealerDocument(dealerOrId);
        dealer.vehicles.pull(vehicle);
        await dealer.save();
    }

    async setVehicles(
        dealerOrId: DealerDocument | mongoose.Types.ObjectId,
        vehiclesToSet: Array<VehicleDocument | mongoose.Types.ObjectId>,
        updateOppositeSide = false
    ): Promise<void> {
        const dealer: DealerDocument = await this.getDealerDocument(dealerOrId);
        const vehiclesToAdd: Array<mongoose.Types.ObjectId | VehicleDocument> = [];
        // First we check if array contains new vehicles
        for (const vehicleToSet of vehiclesToSet) {
            let hasVehicle = false;
            for (const vehicle of dealer.vehicles) {
                if (vehicleToSet._id.equals(vehicle._id)) {
                    hasVehicle = true;
                    break;
                }
            }
            if (!hasVehicle) {
                vehiclesToAdd.push(vehicleToSet._id);
            }
        }
        // Then we check if any vehicle would be deleted from current list of vehicles
        for (const vehicle of dealer.vehicles) {
            let hasVehicle = false;
            for (const vehicleToSet of vehiclesToSet) {
                if (vehicleToSet._id.equals(vehicle._id)) {
                    hasVehicle = true;
                    break;
                }
            }
            if (!hasVehicle) {
                throw new ConstraintError(`Impossible to delete vehicle ${vehicle._id.toString()} from the dealer, as it wouldn't belong to a dealer anymore.`);
            }
        }
        dealer.vehicles.addToSet(...vehiclesToAdd);
        await dealer.save();
        if (updateOppositeSide) {
            for (const vehicle of vehiclesToAdd) {
                await this.vehicleService.setDealer(vehicle, dealer, false);
            }
        }
    }

    async addVehicles(
        dealerOrId: DealerDocument | mongoose.Types.ObjectId,
        vehicles: Array<VehicleDocument | mongoose.Types.ObjectId>,
        updateOppositeSide = false
    ): Promise<void> {
        const dealer = await this.getDealerDocument(dealerOrId);
        const ids: mongoose.Types.ObjectId[] = [];
        for (const vehicle of vehicles) {
            if (vehicle instanceof mongoose.Types.ObjectId) {
                ids.push(vehicle);
            } else {
                ids.push(vehicle._id);
            }
        }
        dealer.vehicles.addToSet(ids);
        await dealer.save();
        if (updateOppositeSide) {
            for (const vehicle of vehicles) {
                await this.vehicleService.setDealer(vehicle, dealer, false);
            }
        }
    }

    async addVehicle(
        dealerOrId: DealerDocument | mongoose.Types.ObjectId,
        vehicle: VehicleDocument | mongoose.Types.ObjectId,
        updateOppositeSide = false
    ): Promise<void> {
        await this.addVehicles(dealerOrId, [vehicle], updateOppositeSide);
    }

    async getDealerDocument(dealerOrDealerId: DealerDocument | mongoose.Types.ObjectId): Promise<DealerDocument> {
        if (dealerOrDealerId instanceof mongoose.Types.ObjectId) {
            const newDealer = await Dealer.findOne({ _id: dealerOrDealerId }).exec();
            if (newDealer == null) {
                throw new EntityDoesNotExistError(`Dealer with id ${dealerOrDealerId.toString()} does not exist.`);
            }
            return newDealer;
        } else {
            return dealerOrDealerId;
        }
    }
}
