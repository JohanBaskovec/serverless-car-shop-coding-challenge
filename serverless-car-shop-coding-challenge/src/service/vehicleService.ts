import { Vehicle } from '../model/vehicle.js';
import { type CreateVehicleDTO, type UpdateVehicleDTO } from '../model/dto/vehicleDTO.js';
import { EntityDoesNotExistError, MissingKeyInJSONBodyError } from '../error/errors.js';
import { type DealerService } from './dealerService.js';
import { type DealerDocument, type VehicleDocument } from '../generated/mongoose.gen.js';
import mongoose from 'mongoose';
import { type CrudService } from './crudService.js';
import { type APIGatewayProxyEventQueryStringParameters } from 'aws-lambda/trigger/api-gateway-proxy.js';
import { type DeleteResponse } from '../model/response.js';

export class VehicleService implements CrudService<VehicleDocument, CreateVehicleDTO, UpdateVehicleDTO> {
    private _dealerService: DealerService;

    get dealerService(): DealerService {
        return this._dealerService;
    }

    set dealerService(value: DealerService) {
        this._dealerService = value;
    }

    async getAll(queryStringParameters: APIGatewayProxyEventQueryStringParameters | null = null): Promise<VehicleDocument[]> {
        const params: any = {};
        if (queryStringParameters != null) {
            if (queryStringParameters.dealer != null) {
                params.dealer = queryStringParameters.dealer;
            }
        }
        return await Vehicle.find(params).exec();
    }

    async create(dto: CreateVehicleDTO): Promise<VehicleDocument> {
        const vehicle = await new Vehicle(dto).save();
        await this.dealerService.addVehicle(vehicle.dealer, vehicle, false);
        return vehicle;
    }

    async update(dto: UpdateVehicleDTO): Promise<VehicleDocument> {
        if (dto._id == null) {
            throw new MissingKeyInJSONBodyError(['_id']);
        }
        const vehicle = await Vehicle.findOne({ _id: dto._id }).exec();
        if (vehicle == null) {
            throw new EntityDoesNotExistError(`Vehicle with id ${dto._id} does not exist.`);
        }
        if (dto.dealer != null) {
            await this.setDealer(vehicle, new mongoose.Types.ObjectId(dto.dealer), true);
        }
        delete dto.dealer;
        const newVehicle = await Vehicle.findOneAndUpdate({ _id: dto._id }, dto, { new: true, runValidators: true }).exec();
        if (newVehicle == null) {
            throw new EntityDoesNotExistError(`Vehicle with id ${dto._id} does not exist.`);
        }
        return newVehicle;
    }

    async remove(id: string): Promise<DeleteResponse> {
        const vehicle = await Vehicle.findOne({ _id: id }).exec();
        if (vehicle == null) {
            throw new EntityDoesNotExistError(`Vehicle with id ${id} does not exist.`);
        }
        await this.dealerService.removeVehicle(vehicle.dealer, vehicle);
        return await Vehicle.deleteOne({ _id: id }).exec();
    }

    private async getVehicleDocument(vehicleOrVehicleId: VehicleDocument | mongoose.Types.ObjectId | string): Promise<VehicleDocument> {
        if (vehicleOrVehicleId instanceof mongoose.Types.ObjectId || typeof vehicleOrVehicleId === 'string') {
            const newVehicle = await Vehicle.findOne({ _id: vehicleOrVehicleId }).exec();
            if (newVehicle == null) {
                throw new EntityDoesNotExistError(`Vehicle with id ${vehicleOrVehicleId.toString()} does not exist.`);
            }
            return newVehicle;
        } else {
            return vehicleOrVehicleId;
        }
    }

    async setDealer(
        vehicleOrId: VehicleDocument | mongoose.Types.ObjectId,
        dealerOrId: DealerDocument | mongoose.Types.ObjectId,
        updateOppositeSide = false
    ): Promise<void> {
        const vehicle = await this.getVehicleDocument(vehicleOrId);

        await this.dealerService.removeVehicle(vehicle.dealer, vehicle);
        if (updateOppositeSide) {
            await this.dealerService.addVehicle(dealerOrId, vehicle, false);
        }

        vehicle.dealer = dealerOrId;
        await vehicle.save();
    }
}
