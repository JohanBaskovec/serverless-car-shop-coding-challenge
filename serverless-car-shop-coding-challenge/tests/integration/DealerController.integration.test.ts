import { Dealer } from '../../src/model/dealer.js';
import { connectToMongo } from '../../src/model/mongoose-db.js';
import { initializeEnvironmentVariables } from '../../src/utils/configuration.js';
import { type DealerDocument } from '../../src/generated/mongoose.gen.js';
import mongoose from 'mongoose';
import { DealerController } from '../../src/controller/dealerController.js';
import { DealerService } from '../../src/service/dealerService.js';
import { type CreateDealerDTO, type UpdateDealerDTO } from '../../src/model/dto/dealerDTO.js';
import { resetDatabase } from './fakeData.js';
import { assertNotNull } from '../utils/assert.js';
import { Vehicle } from '../../src/model/vehicle.js';
import { VehicleService } from '../../src/service/vehicleService.js';
import {
    asJavascriptArray,
    expectMongoArrayNotToContainId,
    expectMongoArrayToContainId,
    getObjectIdAsString
} from './utils.js';

beforeAll(async() => {
    initializeEnvironmentVariables();
    await connectToMongo();
});

afterAll(async() => {
    await mongoose.disconnect();
});

function compareDealers(a: CreateDealerDTO | DealerDocument | null, b: CreateDealerDTO | DealerDocument | null): void {
    a = assertNotNull(a);
    b = assertNotNull(b);
    expect(a.name).toEqual(b.name);
}

const dealerService = new DealerService();
const vehicleService = new VehicleService();
dealerService.vehicleService = vehicleService;
vehicleService.dealerService = dealerService;
const dealerController = new DealerController(dealerService);

beforeEach(async() => {
    await resetDatabase();
});

describe('DealerController', () => {
    describe('getAllDealers()', () => {
        test('should return an empty array if there are no dealers', async() => {
            await Dealer.deleteMany({}).exec();
            const response = await dealerController.getAll();
            expect(response).toMatchSnapshot();
        });

        test('should return dealers when there are dealers', async() => {
            const response = await dealerController.getAll();
            expect(response).toMatchSnapshot();
        });
    });

    describe('createDealer()', () => {
        test('should create a dealer when the request is correct', async() => {
            // We create a new dealer and move a vehicle from an existing dealer to it
            const previousDealer = assertNotNull(await Dealer.findOne({}).exec());
            const vehicle = assertNotNull(await Vehicle.findOne({ dealer: previousDealer._id }).exec());
            const dto: CreateDealerDTO = {
                name: 'name1',
                vehicles: [vehicle._id.toString()]
            };
            const timestampBeforeRequest = +Date.now();
            const response = await dealerController.create(JSON.stringify(dto));
            const timeStampAfterRequest = +Date.now();
            expect(response.statusCode).toBe(200);

            const dealerResponse = JSON.parse(response.body).data;
            compareDealers(dealerResponse, dto);
            const creationDateTimestamp = +new Date(dealerResponse.creationDate);
            const lastUpdateTimestamp = +new Date(dealerResponse.lastUpdateDate);
            expect(creationDateTimestamp).toBeGreaterThanOrEqual(timestampBeforeRequest);
            expect(creationDateTimestamp).toBeLessThanOrEqual(timeStampAfterRequest);
            expect(lastUpdateTimestamp).toBeGreaterThanOrEqual(timestampBeforeRequest);
            expect(lastUpdateTimestamp).toBeLessThanOrEqual(timeStampAfterRequest);

            const dealerInDb = assertNotNull(await Dealer.findOne({ _id: dealerResponse._id }).exec());
            compareDealers(dto, dealerInDb);
            expectMongoArrayToContainId(dealerInDb.vehicles, vehicle);
            const previousDealerAfterUpdate = assertNotNull(await Dealer.findOne({ _id: previousDealer._id }).exec());
            expectMongoArrayNotToContainId(previousDealerAfterUpdate.vehicles, vehicle);

            const vehicleAfter = assertNotNull(await Vehicle.findOne({ _id: vehicle._id }).exec());
            expect(getObjectIdAsString(vehicleAfter.dealer)).toBe(dealerInDb._id.toString());
        });

        test('should return an error when there are missing fields', async() => {
            const dto: any = {};
            const response = await dealerController.create(JSON.stringify(dto));
            expect(response).toMatchSnapshot();
        });

        test('should return an error when using string type for vehicles', async() => {
            const dto: any = { vehicles: '' };
            const response = await dealerController.create(JSON.stringify(dto));
            expect(response).toMatchSnapshot();
        });

        test('should return an error when using an invalid vehicles id', async() => {
            const dto: any = {
                vehicles: ['not-an-id']
            };
            const response = await dealerController.create(JSON.stringify(dto));
            expect(response).toMatchSnapshot();
        });
    });

    describe('removeDealer()', () => {
        test('should remove the dealer when the request is correct', async() => {
            const dealerToRemove = await new Dealer({ name: 'dealer!' }).save();
            const nDealersBefore = await Dealer.count().exec();
            const response = await dealerController.remove({ id: dealerToRemove._id.toString() });
            const nDealersAfter = await Dealer.count().exec();
            expect(response).toMatchSnapshot();

            const dealerInDb = await Dealer.findOne({ _id: dealerToRemove._id }).exec();
            expect(dealerInDb).toBeNull();
            expect(nDealersBefore).toBe(nDealersAfter + 1);
        });

        test('should return an error when trying to delete a dealer that has vehicles', async() => {
            // a Vehicle MUST HAVE a dealer, therefore it's an error to delete a Dealer that has vehicles
            const nDealersBefore = await Dealer.count().exec();
            const dealerWithVehicles = assertNotNull(await Dealer.findOne({ vehicles: { $exists: true, $ne: [] } }).exec());
            const response = await dealerController.remove({ id: dealerWithVehicles._id.toString() });
            const nDealersAfter = await Dealer.count().exec();
            expect(response.statusCode).toBe(400);
            expect(response).toMatchSnapshot();
            expect(nDealersBefore).toBe(nDealersAfter);
        });

        test('should return an error when no id provided', async() => {
            const nDealersBefore = await Dealer.count().exec();
            const response = await dealerController.remove({});
            const nDealersAfter = await Dealer.count().exec();
            expect(response).toMatchSnapshot();
            expect(nDealersBefore).toBe(nDealersAfter);
        });

        test('should return an error when dealer with id does not exist', async() => {
            const id = '754171318ce5b4f5098621cb';
            const nDealersBefore = await Dealer.count().exec();
            const response = await dealerController.remove({ id });
            const nDealersAfter = await Dealer.count().exec();
            expect(response).toMatchSnapshot();
            expect(nDealersBefore).toBe(nDealersAfter);
        });
    });

    describe('updateDealer()', () => {
        test("should update the dealer when the request is correct, and the vehicle that's been transferred from another dealer ", async() => {
            const dealerToUpdate = assertNotNull(await Dealer.findOne({}).exec());
            const vehicleOfOtherDealer = assertNotNull(await Vehicle.findOne({ dealer: { $ne: dealerToUpdate._id } }).exec());
            const dto: UpdateDealerDTO = {
                _id: dealerToUpdate._id.toString(),
                name: 'super name!',
                vehicles: [...asJavascriptArray(dealerToUpdate.vehicles).map(v => v._id.toString()), vehicleOfOtherDealer._id.toString()]
            };
            const timestampBeforeRequest = +Date.now();

            const response = await dealerController.update(JSON.stringify(dto));

            expect(response.statusCode).toBe(200);
            const timeStampAfterRequest = +Date.now();
            const dealerResponse = JSON.parse(response.body).data;
            expect(dealerResponse.name).toBe(dto.name);

            const dealerInDb = assertNotNull(await Dealer.findOne({ _id: dealerToUpdate._id }).exec());
            compareDealers(dealerInDb, dealerResponse);
            const creationTimestamp = +assertNotNull(dealerInDb.creationDate);
            expect(creationTimestamp).toBeLessThan(timestampBeforeRequest);
            const lastUpdateTimestamp = +assertNotNull(dealerInDb.lastUpdateDate);
            expect(lastUpdateTimestamp).toBeGreaterThanOrEqual(timestampBeforeRequest);
            expect(lastUpdateTimestamp).toBeLessThanOrEqual(timeStampAfterRequest);
            expect(dealerInDb.name).toBe(dto.name);

            expectMongoArrayToContainId(dealerInDb.vehicles, vehicleOfOtherDealer);
            const previousDealer = assertNotNull(await Dealer.findOne({ _id: vehicleOfOtherDealer.dealer }).exec());
            expectMongoArrayNotToContainId(previousDealer.vehicles, vehicleOfOtherDealer);

            const vehicleAfter = assertNotNull(await Vehicle.findOne({ _id: vehicleOfOtherDealer._id }).exec());
            expect(getObjectIdAsString(vehicleAfter.dealer)).toBe(dealerInDb._id.toString());
        });

        test('should return an error when removing vehicles', async() => {
            // Vehicles should always have a dealer, therefore it's an error to remove a vehicle from a dealer,
            // because we don't know which dealer the vehicle should belong to
            const dealer = assertNotNull(await Dealer.findOne({ _id: '644171318ce5b4f5098621cb' }).exec());
            const dto: any = {
                _id: dealer._id,
                vehicles: ['644170e85d3ab81e12fcca7e']
            };
            const response = await dealerController.update(JSON.stringify(dto));
            expect(response).toMatchSnapshot();
        });

        test('should return an error when dealer with id does not exist', async() => {
            const dealerToUpdate = assertNotNull(await Dealer.findOne({}).exec());
            const dto: UpdateDealerDTO = {
                _id: '754171318ce5b4f5098621cb',
                name: 'super name!'
            };
            const response = await dealerController.update(JSON.stringify(dto));
            expect(response).toMatchSnapshot();

            const dealerInDb = assertNotNull(await Dealer.findOne({ _id: dealerToUpdate._id }).exec());
            expect(dealerInDb.name).toBe(dealerToUpdate.name);
        });
    });
});
