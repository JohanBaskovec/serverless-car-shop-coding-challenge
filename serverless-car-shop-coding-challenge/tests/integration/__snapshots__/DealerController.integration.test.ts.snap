// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DealerController createDealer() should return an error when there are missing fields 1`] = `
{
  "body": "{"error":{"type":"ValidationError","message":"Dealer validation failed: name: Path \`name\` is required."}}",
  "statusCode": 400,
}
`;

exports[`DealerController createDealer() should return an error when using an invalid vehicles id 1`] = `
{
  "body": "{"error":{"type":"ValidationError","message":"Dealer validation failed: vehicles.0: Cast to [ObjectId] failed for value \\"[ 'not-an-id' ]\\" (type string) at path \\"vehicles.0\\" because of \\"CastError\\", name: Path \`name\` is required."}}",
  "statusCode": 400,
}
`;

exports[`DealerController createDealer() should return an error when using string type for vehicles 1`] = `
{
  "body": "{"error":{"type":"ValidationError","message":"Dealer validation failed: vehicles.0: Cast to [ObjectId] failed for value \\"[ '' ]\\" (type string) at path \\"vehicles.0\\" because of \\"CastError\\", name: Path \`name\` is required."}}",
  "statusCode": 400,
}
`;

exports[`DealerController getAllDealers() should return an empty array if there are no dealers 1`] = `
{
  "body": "{"data":[]}",
  "statusCode": 200,
}
`;

exports[`DealerController getAllDealers() should return dealers when there are dealers 1`] = `
{
  "body": "{"data":[{"_id":"644171318ce5b4f5098621cb","name":"Dealer 1","vehicles":["644170e85d3ab81e12fcca7c","644170e85d3ab81e12fcca7e"],"creationDate":"2009-10-10T00:00:00.000Z","lastUpdateDate":"2009-10-10T00:00:00.000Z","__v":1},{"_id":"644171318ce5b4f5098621ca","name":"Dealer 2","vehicles":["644170e85d3ab81e12fcca7f"],"creationDate":"2008-10-10T00:00:00.000Z","lastUpdateDate":"2008-10-10T00:00:00.000Z","__v":1}]}",
  "statusCode": 200,
}
`;

exports[`DealerController removeDealer() should remove the dealer when the request is correct 1`] = `
{
  "body": "{"data":{"acknowledged":true,"deletedCount":1}}",
  "statusCode": 200,
}
`;

exports[`DealerController removeDealer() should return an error when dealer with id does not exist 1`] = `
{
  "body": "{"error":{"type":"EntityDoesNotExistError","message":"Dealer with id 754171318ce5b4f5098621cb does not exist."}}",
  "statusCode": 404,
}
`;

exports[`DealerController removeDealer() should return an error when no id provided 1`] = `
{
  "body": "{"error":{"type":"MissingPathParametersError","message":"Missing path parameters: id"}}",
  "statusCode": 400,
}
`;

exports[`DealerController removeDealer() should return an error when trying to delete a dealer that has vehicles 1`] = `
{
  "body": "{"error":{"type":"ConstraintError","message":"Dealer has vehicles, delete them first before deleting the dealer."}}",
  "statusCode": 400,
}
`;

exports[`DealerController updateDealer() should return an error when dealer with id does not exist 1`] = `
{
  "body": "{"error":{"type":"EntityDoesNotExistError","message":"Dealer with id 754171318ce5b4f5098621cb does not exist."}}",
  "statusCode": 404,
}
`;

exports[`DealerController updateDealer() should return an error when removing vehicles 1`] = `
{
  "body": "{"error":{"type":"ConstraintError","message":"Impossible to delete vehicle 644170e85d3ab81e12fcca7c from the dealer, as it wouldn't belong to a dealer anymore."}}",
  "statusCode": 400,
}
`;
